### PR TITLE
feat: implement search feature and GUI to handler's archive (hl-684)

### PR DESCRIPTION
**Pull request by sirtawast**
## Description :sparkles:

Add a powerful search feature to find needles from haystack.

* Find SSN by exact match
* Find Ahjo case ID and application number by substring
* Find company by trigram search
* Filter first / last name pairs by using in-memory fuzzy match (Levenshtein)
  * Can be combined with finding by company to improve search time 

## Install trigram functions to PostgreSQL

Migration is done in a file utilizing `TrigramExtension` class.

After migration, you should be able to perform: 

```
from django.contrib.postgres.search import (
    SearchQuery,
    SearchRank,
    SearchVector,
    TrigramSimilarity,
)

search_string = "vincit" # or whatever you want to find
query = SearchQuery(search_string, search_type="websearch")
search_vectors = SearchVector("company__name", weight="A")
Application.objects.filter(archived=False).annotate(search=search_vectors,similarity=TrigramSimilarity("company__name", search_string),rank=SearchRank(search_vectors,query)).filter(similarity__gt=0.3).order_by("-similarity", "-handled_at")
```

### CODE REVIEW

**Comment by terovirtanen** [comment]
**HANDLER branch is deployed to platta: https://helsinkilisa-ui-handler-pr3039.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**APPLICANT branch is deployed to platta: https://helsinkilisa-ui-pr3039.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**API branch is deployed to platta: https://helsinkilisa-pr3039.api.dev.hel.ninja/healthz** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## TestCafe result is __success__ for https://helsinkilisa-ui-pr3039.dev.hel.ninja :laughing::tada::tada::tada:

**Comment by terovirtanen** [comment]
## TestCafe handler result is __failed__ for https://helsinkilisa-ui-handler-pr3039.dev.hel.ninja :crying_cat_face::anger::boom::boom:

**Comment by terovirtanen** [comment]
**API branch is deployed to platta: https://helsinkilisa-pr3039.api.dev.hel.ninja/healthz** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**HANDLER branch is deployed to platta: https://helsinkilisa-ui-handler-pr3039.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**APPLICANT branch is deployed to platta: https://helsinkilisa-ui-pr3039.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## TestCafe result is __success__ for https://helsinkilisa-ui-handler-pr3039.dev.hel.ninja :laughing::tada::tada::tada:

**Comment by terovirtanen** [comment]
**API branch is deployed to platta: https://helsinkilisa-pr3039.api.dev.hel.ninja/healthz** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**APPLICANT branch is deployed to platta: https://helsinkilisa-ui-pr3039.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## TestCafe result is __success__ for https://helsinkilisa-ui-pr3039.dev.hel.ninja :laughing::tada::tada::tada:

**Comment by terovirtanen** [comment]
## TestCafe result is __success__ for https://helsinkilisa-ui-handler-pr3039.dev.hel.ninja :laughing::tada::tada::tada:

File: frontend/benefit/handler/src/components/applicationsArchive/ApplicationArchiveList.tsx [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+          <LoadingSkeleton
+            borderRadius={0}
+            baseColor={theme.colors.fog}
+            height={50}
+          />
+          <LoadingSkeleton height={67} />
+        </>
+      )}
+
+      {hasSearchLoadedWithResults && (
```
Comment (EmiliaMakelaVincit) [inline frontend/benefit/handler/src/components/applicationsArchive/ApplicationArchiveList.tsx:146]: I wonder if the following situation could be made a bit less confusing:
1) Initiate any search, press enter (search results appear)
2) Clear the search (full list appears)
3) Start typing (previous results appear again until a new search is submitted)

Functionally that's not really different from modifying the existing search, but moving from the cleared list back to unrelated results might feel more jarring. The results could maybe be cleared when the search text is cleared.

File: frontend/benefit/handler/public/locales/fi/common.json [hunk 2]
```diff
@@ -1598,12 +1599,14 @@
...
-        "label": "Suodata",
-        "placeholder": "Avainsana"
+      "keyword": {
+        "helperText": "Kirjoita hakemusnumero, työllistettävän nimi tai hetu, työnantajan nimi, y-tunnus tai diaari",
+        "label": "Etsi hakemusta",
+        "placeholder": "Kirjoita hakusana"
       }
     },
-    "keyword": "Hakusana"
+    "perfomingSearch": "Etsitään ...",
```
Comment (EmiliaMakelaVincit) [inline frontend/benefit/handler/public/locales/fi/common.json:1608]: There should be no space before the ellipsis.

**Review by EmiliaMakelaVincit — APPROVED** [review]
Looks good in general, so I'll approve this. There's a few things I wanted to point out that could be changed, but that can also be left as is on your discretion; you can either merge this as is or act on them and then merge afterwards.

File: frontend/benefit/handler/src/components/header/useHeader.tsx [hunk 3]
```diff
@@ -43,7 +43,7 @@ const useHeader = (): ExtendedComponentProps => {
         label: (
           <>
             {t('common:header.navigation.alterations')}
-            {!isAlterationListLoading && alterationData.length > 0 && (
+            {!isAlterationListLoading && alterationData?.length > 0 && (
```
[the remark this answers is not in the document]
Comment (sirtawast) [reply frontend/benefit/handler/src/components/header/useHeader.tsx:46]: Yep, I think you fixed this earlier it but I just applied my fix rebased as well.

[the remark this answers is not in the document]
Comment (sirtawast) [reply frontend/benefit/handler/src/components/header/useHeader.tsx:46]: Yep, I think you fixed this earlier it but I just applied my fix rebased as well.

**Comment by terovirtanen** [comment]
**API branch is deployed to platta: https://tet-pr3039.api.dev.hel.ninja/healthz** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**API branch is deployed to platta: https://kesaseteli-pr3039.api.dev.hel.ninja/healthz** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**ADMIN branch is deployed to platta: https://tet-admin-ui-pr3039.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**EMPLOYER branch is deployed to platta: https://kesaseteli-employer-pr3039.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## TestCafe result is __success__ for https://tet-admin-ui-pr3039.dev.hel.ninja :laughing::tada::tada::tada:

**Comment by terovirtanen** [comment]
**APPLICANT branch is deployed to platta: https://helsinkilisa-ui-pr3039.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**API branch is deployed to platta: https://helsinkilisa-pr3039.api.dev.hel.ninja/healthz** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**HANDLER branch is deployed to platta: https://helsinkilisa-ui-handler-pr3039.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## TestCafe employer result is __failed__ for https://kesaseteli-employer-pr3039.dev.hel.ninja :crying_cat_face::anger::boom::boom:

**Comment by terovirtanen** [comment]
## TestCafe result is __success__ for https://helsinkilisa-ui-handler-pr3039.dev.hel.ninja :laughing::tada::tada::tada:

**Comment by terovirtanen** [comment]
**YOUTH branch is deployed to platta: https://tet-youth-ui-pr3039.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
## TestCafe result is __success__ for https://tet-youth-ui-pr3039.dev.hel.ninja :laughing::tada::tada::tada:

**Pull request merged**